imap_processing/ena_maps/ena_maps.py

```diff
@@ -0,0 +1,543 @@
+        """
+        return (
+            f"UltraPointingSet\n\t(spice_reference_frame="
+            f"{self.spice_reference_frame}, epoch={self.epoch}, "
+            f"num_points={self.num_points})"
+        )
+
+
+# Define the Map classes
+class AbstractSkyMap(ABC):
+    """Abstract base class to contain map data in the context of ENA sky maps."""
+
+    @abstractmethod
+    def __init__(self) -> None:
+        pass
+
+    def __repr__(self) -> str:
+        """
+        Return a string representation of the map.
+
+        Returns
+        -------
+        str
+            String representation of the map.
+        """
+        return f"{self.__class__} Map)"
+
+
+class RectangularSkyMap(AbstractSkyMap):
+    """
+    Map which tiles the sky with a 2D rectangular grid of azimuth/elevation pixels.
+
+    NOTE: Internally, the map is stored as a 1D array of pixels.
+
+    Parameters
+    ----------
+    spacing_deg : float
+        The spacing of the rectangular grid in degrees.
+    spice_frame : geometry.SpiceFrame
+        The reference Spice frame of the map.
+    """
+
+    def __init__(
+        self,
+        spacing_deg: float,
+        spice_frame: geometry.SpiceFrame,
+    ):
+        # Define the core properties of the map:
+        self.tiling_type = SkyTilingType.RECTANGULAR  # Type of tiling of the sky
+        self.spacing_deg = spacing_deg
+        self.spice_reference_frame = spice_frame
+        self.sky_grid = spatial_utils.AzElSkyGrid(
+            spacing_deg=self.spacing_deg,
+        )
+
+        # Solid angles of each pixel in the map grid in units of steradians
+        self.solid_angle_grid = spatial_utils.build_solid_angle_map(
+            spacing_deg=self.spacing_deg,
+        )
+
+        # Unwrap the az, el, solid angle grids to series of points tiling the sky
+        az_points = self.sky_grid.az_grid.ravel(order=RAVEL_ORDER)
+        el_points = self.sky_grid.el_grid.ravel(order=RAVEL_ORDER)
+        self.az_el_points = np.column_stack((az_points, el_points))
+        self.solid_angle_points = self.solid_angle_grid.ravel(order=RAVEL_ORDER)
+        self.num_points = self.az_el_points.shape[0]
+
+        # Initialize empty data dictionary to store map data
+        self.data_dict: dict[str, NDArray] = {}
+
+    def project_pset_values_to_map(
+        self,
+        pointing_set: PointingSet,
+        value_keys: list[tuple[str, IndexMatchMethod]] | None = None,
+    ) -> None:
+        """
+        Project a pointing set's values to the map grid.
+
+        Here, the term "project" refers to the process of determining which pixels in
+        the map grid correspond to which pixels in the pointing set grid, and then
+        binning the values at those indices from the pointing set to the map.
+
+        Parameters
+        ----------
+        pointing_set : PointingSet
+            The pointing set containing the values to project to the map.
+        value_keys : list[tuple[str, IndexMatchMethod]] | None
```

Comment:
Do we need the `IndexMatchMethod` anymore? I thought we might be able to get rid of this with the input/output swapping from before.

Reply:
The Map would still need to know whether a given variable should be pushed or pulled (not the same for all values in the case of Rectangular Ultra PSETs). That will determine which order it passes the objects into the transformer function.

Reply:
Can these be separate calls then? One for the Pull data, one for the push data `project_data(xr_dataset[variables], IndexMatchMethod)`

Reply:
Sure!